source/sigv4.c

```diff
@@ -1929,36 +1970,43 @@ static SigV4Status_t generateCredentialScope( const SigV4Parameters_t * pSigV4Pa
                 pBufLoc += encodedLen;
                 remainingLen -= encodedLen;
 
-                /* An empty value corresponds to an empty string. */
-                if( pCanonicalRequest->pQueryLoc[ i ].value.dataLen > 0U )
+                /* Encode parameter value if non-empty. Query parameters can have empty values. */
+                if( pCanonicalRequest->pQueryLoc[ paramsIndex ].value.dataLen > 0U )
                 {
-                    assert( pCanonicalRequest->pQueryLoc[ i ].value.pData != NULL );
+                    assert( pCanonicalRequest->pQueryLoc[ paramsIndex ].value.pData != NULL );
                     returnStatus = writeValueInCanonicalizedQueryString( pBufLoc,
                                                                          remainingLen,
-                                                                         pCanonicalRequest->pQueryLoc[ i ].value.pData,
-                                                                         pCanonicalRequest->pQueryLoc[ i ].value.dataLen,
+                                                                         pCanonicalRequest->pQueryLoc[ paramsIndex ].value.pData,
+                                                                         pCanonicalRequest->pQueryLoc[ paramsIndex ].value.dataLen,
                                                                          &encodedLen );
                     pBufLoc += encodedLen;
                     remainingLen -= encodedLen;
                 }
             }
 
-            if( ( remainingLen < 1U ) && ( numberOfParameters != ( i + 1U ) ) )
+            /* Check if there is a next parameter. */
+            thereExistsNextParameter = ( ( paramsIndex + 1U ) != numberOfParameters );
+
+            /* Before adding the '&' for the next query parameter, check that this is not
+             * the last parameter. */
+            if( thereExistsNextParameter && ( remainingLen > 0U ) )
             {
-                returnStatus = SigV4InsufficientMemory;
-                LOG_INSUFFICIENT_MEMORY_ERROR( "write the canonical query", 1U );
+                *pBufLoc = '&';
+                ++pBufLoc;
+                remainingLen--;
             }
 
-            if( returnStatus != SigV4Success )
+            /* There is at least another query parameter entry to encode
+             * but no space in the buffer. */
+            else if( thereExistsNextParameter )
             {
+                returnStatus = SigV4InsufficientMemory;
+                LogError( ( "Unable to write canonical query: Insufficient memory configured in \"SIGV4_PROCESSING_BUFFER_LENGTH\"" ) );
```

Comment:
Why enclose this in quotation marks? Better to refer to it as a macro and to mention about how much memory is lacking.